pilot/adapter/config/file/monitor.go

```diff
@@ -0,0 +1,390 @@
+func ParseYamlFile(path string) ([]*model.Config, error) {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return ParseYaml(data)
+}
+
+// ParseYaml is a utility method that parses the contents of a given yaml file and returns all of the Istio configs.
+func ParseYaml(data []byte) ([]*model.Config, error) {
+	resources := parseYaml(data)
+	if len(resources) == 0 {
+		return nil, nil
+	}
+
+	configs := make([]*model.Config, len(resources))
+	for i, r := range resources {
+		var err error
+		if configs[i], err = r.toConfig(); err != nil {
+			return nil, err
+		}
+	}
+
+	return configs, nil
+}
+
+// parseYaml parses the data for an entire yaml file and returns slices for individual resources.
+func parseYaml(data []byte) []*resource {
+	if bytes.HasPrefix(data, []byte("---\n")) {
+		data = data[4:]
+	}
+	if bytes.HasSuffix(data, []byte("\n")) {
+		data = data[:len(data)-1]
+	}
+	if bytes.HasSuffix(data, []byte("\n---")) {
+		data = data[:len(data)-4]
+	}
+	if len(data) == 0 {
+		return nil
+	}
+	chunks := bytes.Split(data, []byte("\n---\n"))
+	resources := make([]*resource, 0, len(chunks))
+	for i, chunk := range chunks {
+		r, err := parseResource(chunk)
+		if err != nil {
+			glog.Errorf("Error processing config [%d]: %v", i, err)
+			continue
+		}
+		if r == nil {
+			continue
+		}
+		resources = append(resources, r)
+	}
+	return resources
+}
+
+// parseResource parses a subsection of yaml for a single resource.
+func parseResource(data []byte) (*resource, error) {
+	r := &resource{}
+	if err := yaml.Unmarshal(data, r); err != nil {
+		return nil, err
+	}
+	if empty(r) {
+		// Can be empty if no data or if it just contains comments.
+		return nil, nil
+	}
+	if r.Kind == "" && r.Metadata.Namespace == "" && r.Metadata.Name == "" {
+		return nil, fmt.Errorf("key elements are empty.\n <<%s>>", string(data))
+	}
+	r.sha = sha1.Sum(data)
+	return r, nil
+}
+
+// Check if the parsed resource is empty
+func empty(r *resource) bool {
+	return reflect.DeepEqual(*r, *emptyResource)
+}
+
+// NewMonitor creates a new config store that monitors files under the given root directory for changes to config.
+// If no kinds are provided (nil or empty), all IstioConfigTypes will be allowed.
+func NewMonitor(delegateStore model.ConfigStore, rootDirectory string, types []string) *Monitor {
+	monitor := &Monitor{
+		store:         delegateStore,
+		root:          rootDirectory,
+		kinds:         map[string]bool{},
+		checkDuration: defaultDuration,
+	}
+
+	if len(types) == 0 {
+		types = model.IstioConfigTypes.Types()
+	}
+
+	for _, k := range types {
+		if schema, ok := model.IstioConfigTypes.GetByType(k); ok {
+			monitor.kinds[messageNameToKind(schema.MessageName)] = true
+		}
+	}
+
+	return monitor
+}
+
+// Start launches a thread that monitors files under the root directory, updating the underlying config
+// store when changes are detected.
+func (m *Monitor) Start(ctx context.Context) {
```

Comment:
Move this up after NewMonitor as they're the public interface exposed by Monitor; same general thinking as in Java class member ordering: public things first then private things, and the constructor is typically the first method in the file.